registry/domain/authentication.js

```diff
@@ -0,0 +1,50 @@
+'use strict';
+
+var path = require('path');
+var strings = require('../../resources/');
+var format = require('stringformat');
+
+var builtin = {
+  basic: {
+    validate: function(authConfig){
+      var isValid = authConfig.username && authConfig.password;
+      return {
+          isValid: isValid,
+          message: isValid ? '' : strings.errors.registry.CONFIGURATION_PUBLISH_BASIC_AUTH_CREDENTIALS_MISSING
+      };
+    },
+    middleware: function(authConfig){
+      var express = require('express');
+      return express.basicAuth(authConfig.username, authConfig.password);
+    }
+  }
+};
+
+var scheme;
+
+module.exports.validate = function(authConfig){
+  if(builtin[authConfig.type]){
+    scheme = builtin[authConfig.type];
+  }
+  else {
+    var cwd = process.cwd();
+    module.paths.push(cwd, path.join(cwd, 'node_modules'));
```

Comment:
It's equivalent to `require.paths.push()`. It basically adds the parent node_modules folder to the list of paths searched by require. So that you can require a module in the upper node_modules folder.

I nicked it from here: https://github.com/mochajs/mocha/blob/master/lib/mocha.js#L28

It's probably not the best way to do it. But I wasn't sure how to proceed in a world where npm3 might or might not flatten the modules folder.

Modifying the paths at runtime is bad, but I figured it's better than using `../../`.